lib/CountdownEvent.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+namespace Amp;
+
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * A countdown event is a promise that is resolved when it was signaled n times
+ *
+ * **Example**
+ *
+ * ```php
+ * $countdownEvent = new CountdownEvent(2);
+ * $countdownEvent->signal();
+ * $countdownEvent->signal(); // promise is now resolved
+ * ```
+ */
+final class CountdownEvent implements Promise
```

Comment:
I don't really like the name event for something that implements Promise. Maybe CountdownBarrier?

Reply:
Barrier isn't the right name either, because that'd mean all `signal()` calls return promises that resolve once the count is 0.

Reply:
A great, I just renamed it to Barrier, haha.
For people coming from .NET CountdownEvent is definitely a suitable name, I don't know what describes it better.